### PR TITLE
Add functions to return raw results for Percentile TDigest and Est

### PR DESCRIPTION
**Pull request by lakshmanan-v**
## Description
Adding aggregate functions to return serialized / raw values of PercentileEst (QuantileDigest) and PercentileTDigest (TDigest) data structures. 
## Upgrade Notes
Does this PR prevent a zero down-time upgrade? (Assume upgrade order: Controller, Broker, Server, Minion)
* [ ] Yes (Please label as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR fix a zero-downtime upgrade introduced earlier?
* [ ] Yes (Please label this as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR otherwise need attention when creating release notes? Things to consider:
- New configuration options
- Deprecation of configurations
- Signature changes to public methods/interfaces
- New plugins added or old plugins removed
* [ ] Yes (Please label this PR as **<code>release-notes</code>** and complete the section on Release Notes)
## Documentation

Will raise an another PR to update the documentation once this PR is approved and the changes are finalized.

### CODE REVIEW

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/PercentileRawEstAggregationFunction.java
Comment (mayankshriv) [inline pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/PercentileRawEstAggregationFunction.java:44]: Add explicit access modifier?

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/PercentileRawEstAggregationFunction.java
Comment (mayankshriv) [inline pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/PercentileRawEstAggregationFunction.java:33]: Please add java docs for all public class/methods?

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/PercentileRawEstAggregationFunction.java
Comment (mayankshriv) [inline pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/PercentileRawEstAggregationFunction.java:135]: Just talking to myself here - we have used this utility function in several places, and I can never remember its name (including ByteUtils). Perhaps we should make a Pinot Utils wrapper and use that. (Not a blocker for this PR).

File: pinot-core/src/test/java/org/apache/pinot/queries/InterSegmentAggregationMultiValueQueriesTest.java
Comment (mayankshriv) [inline pinot-core/src/test/java/org/apache/pinot/queries/InterSegmentAggregationMultiValueQueriesTest.java:523]: We should be able to refactor these tests so that we don't have to write the same code for each of the percentile value test right?

File: pinot-core/src/test/java/org/apache/pinot/queries/InterSegmentAggregationSingleValueQueriesTest.java
Comment (mayankshriv) [inline pinot-core/src/test/java/org/apache/pinot/queries/InterSegmentAggregationSingleValueQueriesTest.java:416]: same here.

File: pinot-core/src/test/java/org/apache/pinot/queries/InterSegmentAggregationMultiValueQueriesTest.java
Comment (lakshmanan-v) [reply pinot-core/src/test/java/org/apache/pinot/queries/InterSegmentAggregationMultiValueQueriesTest.java:523]: Most of the tests in this file use this skeleton. There was already a refactor to use testInterSegmentAggregationResult() which simplified this a bit. We should be able to refactor it further by passing the query, expected values etc.

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/PercentileRawEstAggregationFunction.java
Comment (lakshmanan-v) [reply pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/PercentileRawEstAggregationFunction.java:135]: @mayankshriv Some of these serialized hex strings are little longer (about 3K bytes). Thinking of compressing this using [Snappy](https://github.com/xerial/snappy-java) or [LZ4](https://github.com/lz4/lz4-java). Did a quick test using LZ4 and got the following results. Pinot seems to [use](https://github.com/apache/pinot/blob/b826f2f588bc8d0d05ef466468c2317119ce46cc/pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/compression/LZ4Compressor.java) LZ4 to compress data in segments already. 

Original Size = 3904; Compressed size = 1599; Compression rate = 59.04200819672132 

`  LZ4Factory factory = LZ4Factory.fastestInstance();
    LZ4Compressor compressor = factory.highCompressor(17);
    int maxCompressedLength = compressor.maxCompressedLength(decompressedLength);
    byte[] compressed = new byte[maxCompressedLength];
    int compressedSize = compressor.compress(data, 0, decompressedLength, compressed, 0, maxCompressedLength);
    double ratio = (1 - compressedSize / (double) originalSize) * 100.0;`

Was wondering if we should convert it back to base64 string or just return byte array (lz4 produces byte array). Let me know if you have any other ideas around compressing the results.

File: pinot-core/src/test/java/org/apache/pinot/queries/InterSegmentAggregationMultiValueQueriesTest.java
Comment (lakshmanan-v) [reply pinot-core/src/test/java/org/apache/pinot/queries/InterSegmentAggregationMultiValueQueriesTest.java:523]: Refactored a bit to avoid duplicate code.

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/PercentileRawEstAggregationFunction.java
Comment (mayankshriv) [reply pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/PercentileRawEstAggregationFunction.java:139]: We can do this outside this PR. Perhaps provide a wrapper to encode/decode these bytes, and share it in the client side library. Once we have a wrapper, we can do all sorts of things like compression etc. One reason why we did String was so UI can show it without breaking. Would have to evaluate that as well.

File: pinot-core/src/test/java/org/apache/pinot/queries/InterSegmentAggregationSingleValueQueriesTest.java
Comment (mayankshriv) [inline pinot-core/src/test/java/org/apache/pinot/queries/InterSegmentAggregationSingleValueQueriesTest.java:418]: These are probably brittle tests since we are hardcoding the expected results? We should probably test that results are within a certain error margin of the accurate percentile?

Comment (lakshmanan-v) [reply pinot-core/src/test/java/org/apache/pinot/queries/InterSegmentAggregationSingleValueQueriesTest.java]: Makes sense. Updated.

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/PercentileRawEstAggregationFunction.java
Comment (lakshmanan-v) [reply pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/PercentileRawEstAggregationFunction.java:139]: Sounds good. Do we have a client side library ?

Comment (mayankshriv) [reply pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/PercentileRawEstAggregationFunction.java:139]: We have the `pinot-client` library.

**Comment by lakshmanan-v** [comment]
plz test

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/PercentileRawEstAggregationFunction.java
Comment (lakshmanan-v) [reply pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/PercentileRawEstAggregationFunction.java:139]: Will look into it and add some compress logic if needed after this PR is landed.

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/PercentileRawEstAggregationFunction.java
Comment (lakshmanan-v) [reply pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/PercentileRawEstAggregationFunction.java:37]: Done.

File: pinot-core/src/test/java/org/apache/pinot/queries/InterSegmentAggregationMultiValueQueriesTest.java
Comment (lakshmanan-v) [reply pinot-core/src/test/java/org/apache/pinot/queries/InterSegmentAggregationMultiValueQueriesTest.java:523]: Done.

File: pinot-core/src/test/java/org/apache/pinot/queries/InterSegmentAggregationSingleValueQueriesTest.java
Comment (lakshmanan-v) [reply pinot-core/src/test/java/org/apache/pinot/queries/InterSegmentAggregationSingleValueQueriesTest.java:416]: Done.

File: pinot-core/src/test/java/org/apache/pinot/queries/InterSegmentAggregationMultiValueQueriesTest.java
Comment (mayankshriv) [inline pinot-core/src/test/java/org/apache/pinot/queries/InterSegmentAggregationMultiValueQueriesTest.java:547]: I was thinking we should modify the tests as follows:
1. Run a query with non raw version - get the result
2. Run the same query with raw version - deserialize - get the result
3. Assert that results from 1 and 2 should be exact match.
This will take away any hardcoding as well as needing to check if the results is within a % margin.

File: pinot-core/src/test/java/org/apache/pinot/queries/ExpectedQueryResult.java
Comment (Jackie-Jiang) [inline pinot-core/src/test/java/org/apache/pinot/queries/ExpectedQueryResult.java:22]: Please prefix with `_` for the member variables

File: pinot-core/src/test/java/org/apache/pinot/queries/ExpectedQueryResult.java
Comment (Jackie-Jiang) [inline pinot-core/src/test/java/org/apache/pinot/queries/ExpectedQueryResult.java:30]: Remove `this.`

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/PercentileRawTDigestAggregationFunction.java
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/PercentileRawTDigestAggregationFunction.java:38]: Rename the variable
```suggestion
  private final PercentileTDigestAggregationFunction _percentileTDigestAggregationFunction;
```

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/PercentileRawTDigestAggregationFunction.java
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/PercentileRawTDigestAggregationFunction.java:49]: Rename the parameter
```suggestion
      PercentileTDigestAggregationFunction percentileTDigestAggregationFunction) {
```

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/PercentileRawEstAggregationFunction.java
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/PercentileRawEstAggregationFunction.java:38]: Same for this class

File: pinot-core/src/test/java/org/apache/pinot/queries/InterSegmentAggregationMultiValueQueriesTest.java
Comment (Jackie-Jiang) [reply pinot-core/src/test/java/org/apache/pinot/queries/InterSegmentAggregationMultiValueQueriesTest.java:547]: The result might not be deterministic because of the merging order of the segments though

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/PercentileRawTDigestAggregationFunction.java
Comment (lakshmanan-v) [reply pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/PercentileRawTDigestAggregationFunction.java:49]: Updated.

File: pinot-core/src/test/java/org/apache/pinot/queries/InterSegmentAggregationMultiValueQueriesTest.java
Comment (lakshmanan-v) [reply pinot-core/src/test/java/org/apache/pinot/queries/InterSegmentAggregationMultiValueQueriesTest.java]: Added hard coded checks for group by queries and comparison against original queries for basic query and query with filter.

File: pinot-core/src/test/java/org/apache/pinot/queries/ExpectedQueryResult.java
Comment (lakshmanan-v) [reply pinot-core/src/test/java/org/apache/pinot/queries/ExpectedQueryResult.java]: updated.

File: pinot-core/src/test/java/org/apache/pinot/queries/ExpectedQueryResult.java
Comment (lakshmanan-v) [reply pinot-core/src/test/java/org/apache/pinot/queries/ExpectedQueryResult.java]: updated.

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/PercentileRawTDigestAggregationFunction.java
Comment (lakshmanan-v) [reply pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/PercentileRawTDigestAggregationFunction.java]: updated.

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/PercentileRawEstAggregationFunction.java
Comment (lakshmanan-v) [reply pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/PercentileRawEstAggregationFunction.java]: updated.

File: pinot-common/src/test/java/org/apache/pinot/common/function/AggregationFunctionTypeTest.java
Comment (Jackie-Jiang) [inline pinot-common/src/test/java/org/apache/pinot/common/function/AggregationFunctionTypeTest.java:51]: The code format seems incorrect

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/PercentileRawTDigestAggregationFunction.java
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/PercentileRawTDigestAggregationFunction.java:49]: Still not correct

File: pinot-core/src/test/java/org/apache/pinot/queries/SerializedBytesQueriesTest.java
Comment (Jackie-Jiang) [inline pinot-core/src/test/java/org/apache/pinot/queries/SerializedBytesQueriesTest.java:97]: Don't expose this value outside of this test

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/customobject/SerializedQuantileDigest.java
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/customobject/SerializedQuantileDigest.java:49]: (nit) new line

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/customobject/SerializedQuantileDigest.java
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/customobject/SerializedQuantileDigest.java:41]: Use `Long.compare`

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/customobject/SerializedQuantileDigest.java
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/customobject/SerializedQuantileDigest.java:35]: Directly save `percentile / 100.0` to avoid repeating computation in `compareTo`

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/customobject/SerializedTDigest.java
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/customobject/SerializedTDigest.java:36]: Same

**Review by Jackie-Jiang — COMMENTED** [review]
We recently enforced the format. Can you please use the [Pinot Style](https://docs.pinot.apache.org/developers/developers-and-contributors/code-setup#setup-ide) to format the code?

File: pinot-core/src/test/java/org/apache/pinot/queries/SerializedBytesQueriesTest.java
Comment (lakshmanan-v) [reply pinot-core/src/test/java/org/apache/pinot/queries/SerializedBytesQueriesTest.java]: Updated.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/customobject/SerializedQuantileDigest.java
Comment (lakshmanan-v) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/customobject/SerializedQuantileDigest.java]: Fixed.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/customobject/SerializedQuantileDigest.java
Comment (lakshmanan-v) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/customobject/SerializedQuantileDigest.java]: Done.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/customobject/SerializedTDigest.java
Comment (lakshmanan-v) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/customobject/SerializedTDigest.java]: Done.

File: pinot-common/src/test/java/org/apache/pinot/common/function/AggregationFunctionTypeTest.java
Comment (lakshmanan-v) [reply pinot-common/src/test/java/org/apache/pinot/common/function/AggregationFunctionTypeTest.java]: Updated.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/customobject/SerializedQuantileDigest.java
Comment (lakshmanan-v) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/customobject/SerializedQuantileDigest.java]: Updated.

**Review by mayankshriv — APPROVED** [review]
Overall looks good, thanks for contributing this. Let's also wait for @Jackie-Jiang 's approval.

**Review by Jackie-Jiang — APPROVED** [review]

**Pull request merged**
